src/api-binder/report.ts

```diff
@@ -141,16 +147,40 @@ export async function startReporting() {
 	])) as StateReportOpts;
 
 	let reportPending = false;
-	const doReport = async () => {
+	const doReport = async (): boolean => {
 		if (!reportPending) {
 			reportPending = true;
 			await reportCurrentState(reportConfigs);
 			reportPending = false;
+			return true;
+		} else {
+			return false;
 		}
 	};
 
+	// Managing report frequency: We can't control the frequency of state change
+	// events or the arrival time of those events relative to scheduled metrics
+	// reports. So, we employ a two layer strategy to manage frequency:
+	// 1) Ensure that below we send reports from the same source (event/scheduled)
+	//    no sooner than the max frequency of that type.
+	// 2) Rely on reportCurrentState() also to validate these frequencies because
+	//    it is the single handler of requests from both sources.
+
+	const onStateChange = async () => {
+		// If report already in progress, it's possible the state change that
+		// triggered this event might be missed, so try once more as best effort.
+		// No worries about duplicating report content, as we expect state content
+		// is diffed to validate a change has occurred. Also, metrics will be
+		// excluded from the follow-up report because maxMetricsFrequency should
+		// be a much larger interval than maxReportFrequency.
+		if (! await doReport()) {
+			await delay(constants.maxReportFrequency);
+			await doReport();
+		}
```

Comment:
I'm not sure once with best effort is enough, given that a change event is emitted for numerous events in quick succession during an update. Practically, how often does current state actually get reported on a device with this PR code?

Reply:
It's still better than before where we tried exactly once, and dropped the report attempt entirely if a report was in progress

Reply:
I have revised the retry mechanism to ensure the most recent device state is reported.